distributed/deploy/spec.py

```diff
@@ -0,0 +1,296 @@
+
+    def _correct_state(self):
+        if self._correct_state_waiting:
+            # If people call this frequently, we only want to run it once
+            return self._correct_state_waiting
+        else:
+            task = asyncio.Task(self._correct_state_internal())
+            self._correct_state_waiting = task
+            return task
+
+    async def _correct_state_internal(self):
+        async with self._lock:
+            self._correct_state_waiting = None
+
+            pre = list(set(self.workers))
+            to_close = set(self.workers) - set(self.worker_spec)
+            if to_close:
+                await self.scheduler.retire_workers(workers=list(to_close))
+                tasks = [self.workers[w].close() for w in to_close]
+                for task in tasks:
+                    await task
+            for name in to_close:
+                del self.workers[name]
+
+            to_open = set(self.worker_spec) - set(self.workers)
+            workers = []
+            for name in to_open:
+                d = self.worker_spec[name]
+                cls, opts = d["cls"], d.get("options", {})
+                if "name" not in opts:
+                    opts = toolz.merge({"name": name}, opts, {"loop": self.loop})
+                worker = cls(self.scheduler.address, **opts)
+                self._created.add(worker)
+                workers.append(worker)
+            if workers:
+                await asyncio.wait(workers)
+                for w in workers:
+                    w._cluster = weakref.ref(self)
+                    if self.status == "running":
+                        await w
+            self.workers.update(dict(zip(to_open, workers)))
+
+    def __await__(self):
+        async def _():
+            if self.status == "created":
+                await self._start()
+            await self.scheduler
+            await self._correct_state()
+            if self.workers:
+                await asyncio.wait(list(self.workers.values()))  # maybe there are more
+            await self._wait_for_workers()
+            return self
+
+        return _().__await__()
+
+    async def _wait_for_workers(self):
+        # TODO: this function needs to query scheduler and worker state
+        # remotely without assuming that they are local
+        while {d["name"] for d in self.scheduler.identity()["workers"].values()} != set(
+            self.workers
+        ):
+            if (
+                any(w.status == "closed" for w in self.workers.values())
+                and self.scheduler.status == "running"
+            ):
+                raise gen.TimeoutError("Worker unexpectedly closed")
+            await asyncio.sleep(0.1)
+
+    async def __aenter__(self):
+        await self
+        return self
+
+    async def __aexit__(self, typ, value, traceback):
+        await self.close()
+
+    async def _close(self):
+        while self.status == "closing":
+            await asyncio.sleep(0.1)
```

Comment:
Instead of polling, could have a future for the closing operation (created by the first call to `_close`), and just wait on that?

Reply:
Good thought.  I'm inclined to wait on this for now though if that's ok.